feat:Added functionality to retrieve and save GitHub token (#812)

wblxxx opened the pull request:
## Brief Information

This pull request is in the type of ([more info](https://github.com/angular/angular/blob/main/CONTRIBUTING.md#type) about types):

- [ ] build
- [ ] ci
- [ ] docs
- [√] feat
- [ ] fix
- [ ] perf
- [ ] refactor
- [ ] test

Related issues ([GitHub tokens](https://github.com/hypertrons/hypertrons-crx/issues/812)):

- [OSS101] Task 8: Obtain GitHub tokens to increase request limits #812

## Details
<!-- What did you do in this PR?  -->
Recently, there has been a https://github.com/hypertrons/hypertrons-crx/issues/773#issuecomment-1918190670 in the community about GitHub Tokens. In fact, HyperCRX had previously introduced GitHub Tokens to increase the number of requests made to GitHub, as seen in https://github.com/hypertrons/hypertrons-crx/pull/100. However, in https://github.com/hypertrons/hypertrons-crx/issues/574, we removed it since there was no longer a frequent need to access the GitHub API.

Considering that HyperCRX may need to frequently access the GitHub API in the future, we need to reintroduce GitHub Tokens with the following specific requirements:

Provide a UI interface allowing users to input their GitHub Tokens.
Securely store user's Tokens.
Encapsulate a network request method that uses the GitHub Token, which will be called by other new features needing to access the GitHub API in the future.

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/hypertrons/hypertrons-crx/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/hypertrons/hypertrons-crx)

## Others
<!-- Other information you want to share.  -->
<!-- If none to share, leave an "N.A."  -->

wxharry commented on src/api/githubApi.ts:
```diff
@@ -0,0 +1,35 @@
+// src/api/githubApi.ts
+let githubToken = '';
+
+export const saveToken = (token: string) => {
+  githubToken = token;
+  localStorage.setItem('github_token', token); // 保存 token 到 localStorage
+};
+
+export const getToken = () => {
+  if (!githubToken) {
+    githubToken = localStorage.getItem('github_token') || '';
+  }
+  return githubToken;
+};
```
save and get token are not api related, could you move them to `helpers/`?

wxharry replied:
I think I meant to move only the saveToken and getToken to helper and keep the githubRequest in the api folder lol

wxharry reviewed (changes requested):
Nice work. We are close. Here are what we can do to improve this branch.
1. Please look into comments and remove or update what are needed to. 
2. We need to use English within the project, so please update all messages to English and use i18next for supporting other languages. reference to #805 
3. Let's move the save button to the same line of the token input form. When user click on save, the input form would turn into a text and show only characters at the beginning and the end, hide the others with `*` and the save button becomes a edit button.

<img width="607" alt="image" src="https://github.com/user-attachments/assets/2a03f619-87bc-444c-b2ec-c1a95ffb62e7">

4. For alerts, we should use user-friendly popup from our ui library. For example, [message from antd](https://ant.design/components/message-cn). Looks like alerts block interactions with browser until we confirm.

5. On test token, can we use the token in the input form? It is intuitive to see the token there and click to test if it is valid. let me know what you think.

cc @HalloMelon for UI changes.

wblxxx commented:
Sorry for the delay in making the changes until today. Thank you for your diligent review. I have resolved all the issues in the latest version. Additionally, using the token in the input form is intuitive and user-friendly, as it enables immediate feedback on the token's validity, which I have adopted. If there are any further issues, please continue to correct me.

wxharry commented on src/pages/Options/Options.tsx:
```diff
@@ -146,10 +149,21 @@ const Options = (): JSX.Element => {
+              alignItems: 'center',
+            }}
+          >
+            <GitHubToken /> {/* 添加 GitHubToken 组件 */}
```
This one is not fixed. Please double check.

wxharry commented on src/helpers/githubApi.ts:
```diff
@@ -0,0 +1,34 @@
+  return githubToken;
+};
+
+export const githubRequest = async (endpoint: string, options: RequestInit = {}) => {
+  const token = getToken();
+  if (!token) {
+    throw new Error('GitHub Token 未设置');
+  }
+
+  const response = await fetch(`https://api.github.com${endpoint}`, {
+    ...options,
+    headers: {
+      ...options.headers,
+      Authorization: `Bearer ${token}`,
+    },
+  });
+
+  if (!response.ok) {
+    throw new Error(`GitHub 请求失败: ${response.statusText}`);
+  }
+
+  return response.json();
+};
```
Can we implement this method in this way:
1. if (!gettoken) return null
2. try fetch and return whatever response and catch and throw unexpected errors.

Something like this: (not tested)
```
export const githubRequest = async (endpoint: string, options: RequestInit = {}) => {
  if (!getToken()) {
    return null
  }

try{
  const response = await fetch(`https://api.github.com${endpoint}`, {
    ...options,
    headers: {
      ...options.headers,
      Authorization: `Bearer ${token}`,
    },
  });
  return response.json();
}
catch{
    return null;
}
};
```

Things I tried to do here:
1. This function only does one thing: make github request.
2. So it only returns response if request succeded or null if no request is made.
3. We don't care about if the response is ok or error in here, whatever code calls this request should handle that. For here, we only make sure the output is valid or null.

Let me know if this makes sense.

wxharry reviewed (changes requested):
Almost there! Some minor enhancements are needed before we merge it. The biggest issue it that I can't see the message, not sure what's happend. I didn't take a close at the code. Let me know if it works for you or you need more info from me.

wxharry commented on src/helpers/githubApi.ts:
```diff
@@ -0,0 +1,34 @@
+  }
+
+  const response = await fetch(`https://api.github.com${endpoint}`, {
+    ...options,
+    headers: {
+      ...options.headers,
+      Authorization: `Bearer ${token}`,
+    },
```
This is causing the bug. options has the same thing as headers, they are duplicate. I think we can just keep `...options` and remove the rest.

wblxxx commented:
Your suggestion makes a lot of sense. I appreciate your effort in reviewing the new version.

wblxxx commented:
I simplified the githubApi.ts and moved the decision logic to GitHubToken.tsx.By the way,if we are only using githubToken to get the token from localStorage, we can directly use localStorage.getItem('github_token') ,the previous approach might reduce the number of accesses to localStorage in scenarios where githubRequest is called frequently, thereby improving performance.

wxharry commented:
> This is causing the bug. options has the same thing as headers, they are duplicate. I think we can just keep `...options` and remove the rest.

Did you solve this? I don't get any response if I set the headers twice. Removing duplicate options works for me but the message is still not showing. Nothing is showing in the console. Do you have any ideas?

wblxxx commented:
> > This is causing the bug. options has the same thing as headers, they are duplicate. I think we can just keep `...options` and remove the rest.
> 
> Did you solve this? I don't get any response if I set the headers twice. Removing duplicate options works for me but the message is still not showing. Nothing is showing in the console. Do you have any ideas?

This is so strange. The day before yesterday, everything was normal when I was using and uploading the code. Yesterday, there were no message  when I first clicked, but during the process of modifying the code, it naturally recovered, and everything remained normal for several hours after uploading. Just now, there were no message when I first clicked, but after having a meal, it was fine. I'm thinking about it.

wblxxx commented:
> > This is causing the bug. options has the same thing as headers, they are duplicate. I think we can just keep `...options` and remove the rest.
> 
> Did you solve this? I don't get any response if I set the headers twice. Removing duplicate options works for me but the message is still not showing. Nothing is showing in the console. Do you have any ideas?

I suspect that the previous fixed position of the message pop-up might have caused it to be invisible in some cases. I have adjusted it to dynamically calculate the position of the message. However, since it has always run successfully on my side, I cannot determine if the bug still exists. Please test it and provide feedback promptly.
![iShot_2024-07-31_15 13 29](https://github.com/user-attachments/assets/0d38655b-3ab0-472d-b6e2-2e17dac0baec)

wblxxx commented:
I checked today, and the issue didn't happen again.

wxharry commented:
> 
> I suspect that the previous fixed position of the message pop-up might have caused it to be invisible in some cases. I have adjusted it to dynamically calculate the position of the message. However, since it has always run successfully on my side, I cannot determine if the bug still exists. Please test it and provide feedback promptly. ![iShot_2024-07-31_15 13 29](https://private-user-images.githubusercontent.com/60059315/353731895-0d38655b-3ab0-472d-b6e2-2e17dac0baec.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI0ODU0MzIsIm5iZiI6MTcyMjQ4NTEzMiwicGF0aCI6Ii82MDA1OTMxNS8zNTM3MzE4OTUtMGQzODY1NWItM2FiMC00NzJkLWI2ZTItMmUxN2RhYzBiYWVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODAxVDA0MDUzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVjZjY4MTJkZmZkYzk4Yzk1MzczYmZkMDBhZjRkZjE2YjNiMTY2Y2M5YTVmM2VjODQ1MjlhZGRkNThhNjA5OWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5UnWtpOBfWLVdEp305XiEKkNE8alpn7Lj17ju6xKrkM)

Would it be easier to move it to the top of the page? because messages usually appear at the top by default. This is not a change request, you can keep it this way if that could cause the message invisible issue.

Also, did we figure out why we have duplicate options here? https://github.com/hypertrons/hypertrons-crx/pull/839#discussion_r1696302272

wblxxx commented:
> > I suspect that the previous fixed position of the message pop-up might have caused it to be invisible in some cases. I have adjusted it to dynamically calculate the position of the message. However, since it has always run successfully on my side, I cannot determine if the bug still exists. Please test it and provide feedback promptly. ![iShot_2024-07-31_15 13 29](https://private-user-images.githubusercontent.com/60059315/353731895-0d38655b-3ab0-472d-b6e2-2e17dac0baec.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI0ODU0MzIsIm5iZiI6MTcyMjQ4NTEzMiwicGF0aCI6Ii82MDA1OTMxNS8zNTM3MzE4OTUtMGQzODY1NWItM2FiMC00NzJkLWI2ZTItMmUxN2RhYzBiYWVjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODAxVDA0MDUzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVjZjY4MTJkZmZkYzk4Yzk1MzczYmZkMDBhZjRkZjE2YjNiMTY2Y2M5YTVmM2VjODQ1MjlhZGRkNThhNjA5OWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5UnWtpOBfWLVdEp305XiEKkNE8alpn7Lj17ju6xKrkM)
> 
> Would it be easier to move it to the top of the page? because messages usually appear at the top by default. This is not a change request, you can keep it this way if that could cause the message invisible issue.
> 
> Also, did we figure out why we have duplicate options here? [#839 (comment)](https://github.com/hypertrons/hypertrons-crx/pull/839#discussion_r1696302272)

By default, the response message appears at the top, but at that time, the user's attention is focused on the GitHubToken component area. If it appears at the top, they might not notice it, and shifting their gaze to focus on the content and then back is also not user-friendly. Therefore, I think the current interaction is better. Additionally, if our function is only to pass in the GitHub token and does not need to retain other request configurations or header information, then ...options and ...options.headers are indeed redundant. The code can be simplified, and I have made the modifications.

wxharry commented:
LGTM. Thanks for your efforts! Looks like the branch needs formatting. Please run `yarn run prettier`.

wblxxx commented:
> LGTM. Thanks for your efforts! Looks like the branch needs formatting. Please run `yarn run prettier`.

I have always used Prettier to check the code format, and there are no issues when checking locally. The PR is also fully synchronized, so I don't know where the problem is.
![iShot_2024-08-02_14 25 53](https://github.com/user-attachments/assets/90eb3ebd-847e-414c-857a-b5676df97ed8)
![iShot_2024-08-02_14 27 11](https://github.com/user-attachments/assets/5631c5ce-a144-448f-bd4c-fdfd1ced74cf)

wxharry commented:
> I have always used Prettier to check the code format, and there are no issues when checking locally. The PR is also fully synchronized, so I don't know where the problem is. ![iShot_2024-08-02_14 25 53](https://private-user-images.githubusercontent.com/60059315/354489650-90eb3ebd-847e-414c-857a-b5676df97ed8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI2NjIyNDEsIm5iZiI6MTcyMjY2MTk0MSwicGF0aCI6Ii82MDA1OTMxNS8zNTQ0ODk2NTAtOTBlYjNlYmQtODQ3ZS00MTRjLTg1N2EtYjU2NzZkZjk3ZWQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODAzVDA1MTIyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU2NzUxMjlhNjlhMzg4MTZmZGU3NTgxZWQwZjUwMDE4MDI1N2I2MjcyNTk5YzU1MjYzMjc3ODFjYmQ2YTIxMDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.bSleSnt9Gi7x6bjx6iyBrdLHAARu4-clciukozwXz3A) ![iShot_2024-08-02_14 28 50](https://private-user-images.githubusercontent.com/60059315/354490259-a0eef8ac-148e-4400-8609-734001381b3f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI2NjIyNDEsIm5iZiI6MTcyMjY2MTk0MSwicGF0aCI6Ii82MDA1OTMxNS8zNTQ0OTAyNTktYTBlZWY4YWMtMTQ4ZS00NDAwLTg2MDktNzM0MDAxMzgxYjNmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODAzVDA1MTIyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQyMmNmZWY2MWY4NTg1NjlkMzRjNGRmNTM3ODlkYzgwMjFkNTk4NDgxNDVjZTQwM2FlNzZkMGM0OGNkMDNjNzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.RCD-mRWK9MZAZh5R1rJPYmuTG--pxdtMopuE_EBPEzs)

This is weird! My prettier gives three files that need to be fixed. Do you mind I adding those changes to this branch?
<img width="782" alt="image" src="https://github.com/user-attachments/assets/f5085da8-6998-4de9-9f02-d914a6cb1556">

wblxxx commented:
> > I have always used Prettier to check the code format, and there are no issues when checking locally. The PR is also fully synchronized, so I don't know where the problem is. ![iShot_2024-08-02_14 25 53](https://private-user-images.githubusercontent.com/60059315/354489650-90eb3ebd-847e-414c-857a-b5676df97ed8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI2NjIyNDEsIm5iZiI6MTcyMjY2MTk0MSwicGF0aCI6Ii82MDA1OTMxNS8zNTQ0ODk2NTAtOTBlYjNlYmQtODQ3ZS00MTRjLTg1N2EtYjU2NzZkZjk3ZWQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODAzVDA1MTIyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU2NzUxMjlhNjlhMzg4MTZmZGU3NTgxZWQwZjUwMDE4MDI1N2I2MjcyNTk5YzU1MjYzMjc3ODFjYmQ2YTIxMDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.bSleSnt9Gi7x6bjx6iyBrdLHAARu4-clciukozwXz3A) ![iShot_2024-08-02_14 28 50](https://private-user-images.githubusercontent.com/60059315/354490259-a0eef8ac-148e-4400-8609-734001381b3f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI2NjIyNDEsIm5iZiI6MTcyMjY2MTk0MSwicGF0aCI6Ii82MDA1OTMxNS8zNTQ0OTAyNTktYTBlZWY4YWMtMTQ4ZS00NDAwLTg2MDktNzM0MDAxMzgxYjNmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODAzVDA1MTIyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQyMmNmZWY2MWY4NTg1NjlkMzRjNGRmNTM3ODlkYzgwMjFkNTk4NDgxNDVjZTQwM2FlNzZkMGM0OGNkMDNjNzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.RCD-mRWK9MZAZh5R1rJPYmuTG--pxdtMopuE_EBPEzs)
> 
> This is weird! My prettier gives three files that need to be fixed. Do you mind I adding those changes to this branch?
> 
> <img alt="image" width="782" src="https://private-user-images.githubusercontent.com/39271899/354796667-f5085da8-6998-4de9-9f02-d914a6cb1556.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI2NjQ1ODcsIm5iZiI6MTcyMjY2NDI4NywicGF0aCI6Ii8zOTI3MTg5OS8zNTQ3OTY2NjctZjUwODVkYTgtNjk5OC00ZGU5LTlmMDItZDkxNGE2Y2IxNTU2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODAzVDA1NTEyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWExN2Q2Yzk1NDZmNzFjNTI2NTcyZWY3NjU3Yjc3ZWEzOGJhYjVhZDdiNjkwY2RmZWI0M2EwZmM5NjA0MzVhMDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Ars6U1fZd6w6ff_QUHExJJ77XWB1extX6cu10OKZCz0">

It looks like the versions are different, but of course, it's okay.
